pkg/utils/apply/apply.go

```diff
@@ -0,0 +1,205 @@
+
+type patcher interface {
+	calculatePatch(c, m runtime.Object) (client.Patch, error)
+}
+
+type patcherFn func(c, m runtime.Object) (client.Patch, error)
+
+func (fn patcherFn) calculatePatch(c, m runtime.Object) (client.Patch, error) {
+	return fn(c, m)
+}
+
+// APIApplicator implements Applicator
+type APIApplicator struct {
+	creator
+	patcher
+	c client.Client
+}
+
+// UpdatingApply applies changes to an object with updating
+func (a *APIApplicator) UpdatingApply(ctx context.Context, desired runtime.Object, ao ...ApplyOption) error {
+	existing, err := a.createOrGetExisting(ctx, a.c, desired, ao...)
+	if err != nil {
+		return err
+	}
+	if existing == nil {
+		return nil
+	}
+
+	// the object already exists, update
+	if err := applyApplyOptions(ctx, existing, desired, ao); err != nil {
+		return err
+	}
+	desired.(Object).SetResourceVersion(existing.(Object).GetResourceVersion())
+	return errors.Wrap(a.c.Update(ctx, desired), "cannot update object")
+}
+
+// PatchingApply applies changes to an object with patching
+func (a *APIApplicator) PatchingApply(ctx context.Context, desired runtime.Object, ao ...ApplyOption) error {
+	existing, err := a.createOrGetExisting(ctx, a.c, desired, ao...)
+	if err != nil {
+		return err
+	}
+	if existing == nil {
+		return nil
+	}
+
+	// the object already exists, patch changes
+	if err := applyApplyOptions(ctx, existing, desired, ao); err != nil {
+		return err
+	}
+	patch, err := a.patcher.calculatePatch(existing, desired)
+	if err != nil {
+		return errors.Wrap(err, "cannot calculate patch by computing a three way diff")
+	}
+	return errors.Wrapf(a.c.Patch(ctx, desired, patch), "cannot patch object")
+}
+
+// createOrGetExisting will create the object if it does not exist
+// or return the existing object if it already exists
+func createOrGetExisting(ctx context.Context, c client.Client, desired runtime.Object, ao ...ApplyOption) (runtime.Object, error) {
+	m, ok := desired.(Object)
+	if !ok {
+		return nil, errors.New("cannot access object metadata")
+	}
+
+	if m.GetName() == "" && m.GetGenerateName() == "" {
+		// apply ApplyOption even cannot get the object by name
+		if err := applyApplyOptions(ctx, nil, desired, ao); err != nil {
+			return nil, err
+		}
+		if err := addLastAppliedStateAnnotation(desired); err != nil {
+			return nil, err
+		}
+		return nil, errors.Wrap(c.Create(ctx, desired), "cannot create object")
+	}
+
+	existing := &unstructured.Unstructured{}
+	existing.GetObjectKind().SetGroupVersionKind(desired.GetObjectKind().GroupVersionKind())
+	err := c.Get(ctx, types.NamespacedName{Name: m.GetName(), Namespace: m.GetNamespace()}, existing)
+	if kerrors.IsNotFound(err) {
+		// apply ApplyOption even though the object does not exist
+		if err := applyApplyOptions(ctx, nil, desired, ao); err != nil {
+			return nil, err
+		}
+		if err := addLastAppliedStateAnnotation(desired); err != nil {
+			return nil, err
+		}
+		return nil, errors.Wrap(c.Create(ctx, desired), "cannot create object")
```

Comment:
BTW, I noticed the code here is exactly the same with L146~L152, if we really need both, we can use an anonymous function for this case.